File: app/routers/webhook.py
from fastapi import APIRouter, Depends, Request
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.order import Order
from app.models.product import Product
from app.services.mercadopago import MercadoPagoService
from app.utils import success_response, error_response

router = APIRouter(prefix="/webhook", tags=["Webhook"])

@router.post("/mercadopago")
async def webhook_mercadopago(request: Request, db: Session = Depends(get_db)):
    try:
        data = await request.json()
        
        if data.get("type") == "payment":
            payment_id = data["data"]["id"]
            
            # Buscar informações do pagamento
            mp_service = MercadoPagoService()
            payment_info = mp_service.get_payment(payment_id)
            
            # Buscar pedido pelo external_reference
            order_id = payment_info.get("external_reference")
            if order_id:
                order = db.query(Order).filter(Order.id == int(order_id)).first()
                if order:
                    # Atualizar status do pedido baseado no status do pagamento
                    payment_status = payment_info.get("status")
                    
                    if payment_status == "approved":
                        order.status = "paid"
                        
                        # Reduzir estoque dos produtos
                        for item in order.items:
                            product = db.query(Product).filter(Product.id == item["product_id"]).first()
                            if product:
                                product.estoque -= item["quantidade"]
                    
                    elif payment_status == "cancelled" or payment_status == "rejected":
                        order.status = "cancelled"
                    
                    db.commit()
        
        return success_response(message="Webhook processado")
    
    except Exception as e:
        return error_response(f"Erro no webhook: {str(e)}", 500)

File: routes/webhook.js
const express = require('express');
const { MercadoPagoConfig, Payment } = require('mercadopago');
const { PrismaClient } = require('@prisma/client');
const { successResponse, errorResponse } = require('../utils/responses');

const router = express.Router();
const prisma = new PrismaClient();

const client = new MercadoPagoConfig({
  accessToken: process.env.MERCADOPAGO_ACCESS_TOKEN
});
const payment = new Payment(client);

// Webhook Mercado Pago
router.post('/mercadopago', async (req, res) => {
  try {
    const { type, data } = req.body;

    if (type === 'payment') {
      const paymentId = data.id;

      // Buscar informações do pagamento
      const paymentInfo = await payment.get({ id: paymentId });
      const orderIdStr = paymentInfo.external_reference;

      if (orderIdStr) {
        const orderId = parseInt(orderIdStr);
        const order = await prisma.order.findUnique({
          where: { id: orderId }
        });

        if (order) {
          const paymentStatus = paymentInfo.status;
          let newStatus = order.status;

          if (paymentStatus === 'approved') {
            newStatus = 'paid';

            // Reduzir estoque dos produtos
            for (const item of order.items) {
              await prisma.product.update({
                where: { id: item.product_id },
                data: {
                  estoque: {
                    decrement: item.quantidade
                  }
                }
              });
            }
          } else if (paymentStatus === 'cancelled' || paymentStatus === 'rejected') {
            newStatus = 'cancelled';
          }

          await prisma.order.update({
            where: { id: orderId },
            data: { status: newStatus }
          });
        }
      }
    }

    return successResponse(res, null, 'Webhook processado');

  } catch (error) {
    console.error(error);
    return errorResponse(res, 'Erro no webhook', 500);
  }
});

module.exports = router;